File: app/src/main/java/com/walhalla/pillfinder/ui/adapter/scroll/PaginationScrollListener.java
package com.walhalla.pillfinder.ui.adapter.scroll;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public abstract class PaginationScrollListener extends RecyclerView.OnScrollListener {

    private final LinearLayoutManager layoutManager;
    private boolean isLoaded;

    public PaginationScrollListener(LinearLayoutManager layoutManager) {
        this.layoutManager = layoutManager;
    }

    @Override
    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        if (isLoaded) {
            int visibleItemCount = layoutManager.getChildCount();
            int totalItemCount = layoutManager.getItemCount();
            int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

            if (!isLoading() && !isLastPage()) {
                if ((visibleItemCount + firstVisibleItemPosition) >=
                        totalItemCount && firstVisibleItemPosition >= 0) {
                    loadMoreItems();
                }
            }

        }
        isLoaded = true;
    }

    protected abstract void loadMoreItems();

    public abstract int getTotalPageCount();

    public abstract boolean isLastPage();

    public abstract boolean isLoading();
}

File: app/src/main/java/com/walhalla/pillfinder/ui/adapter/scroll/PaginationScrollListener.kt
package com.walhalla.pillfinder.ui.adapter.scroll

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

abstract class PaginationScrollListener(private val layoutManager: LinearLayoutManager) : RecyclerView.OnScrollListener() {
    private var isLoaded = false

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        if (isLoaded) {
            val visibleItemCount = layoutManager.childCount
            val totalItemCount = layoutManager.getItemCount()
            val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()

            if (!this.isLoading && !this.isLastPage) {
                if ((visibleItemCount + firstVisibleItemPosition) >=
                    totalItemCount && firstVisibleItemPosition >= 0
                ) {
                    loadMoreItems()
                }
            }
        }
        isLoaded = true
    }

    protected abstract fun loadMoreItems()

    abstract val totalPageCount: Int

    abstract val isLastPage: Boolean

    abstract val isLoading: Boolean
}
